crates/iota-indexer/src/indexer.rs

```diff
@@ -211,3 +337,60 @@ impl ProgressStore for ShimIndexerProgressStore {
         Ok(())
     }
 }
+
+/// Pulls checkpoints from the gRPC stream and sends them to the WorkerPool
+/// channel.
+pub async fn run_grpc_checkpoint_workerpool_ingestion(
+    grpc_url: String,
+    tx: tokio::sync::mpsc::Sender<Arc<CheckpointData>>,
+    start_watermark: u64,
+    cancel: CancellationToken,
+) -> Result<(), IndexerError> {
+    use tokio_stream::StreamExt;
+    let mut client = GrpcNodeClient::connect(&grpc_url)
+        .await
+        .map_err(|e| IndexerError::Generic(format!("Failed to connect to gRPC: {e}")))?;
+    info!(
+        "[gRPC][Indexer] starting stream from watermark {} (WorkerPool mode)",
+        start_watermark
+    );
+    let mut stream = client
+        .stream_checkpoints(Some(start_watermark), None, Some(true))
+        .await
+        .map_err(|e| IndexerError::Generic(format!("Failed to stream checkpoints: {e}")))?;
+    while let Some(result) = stream.next().await {
+        if cancel.is_cancelled() {
+            warn!("[gRPC][Indexer] Cancelled, stopping stream");
+            break;
+        }
```

Comment:
Return here, and at the end of function, we can return Err without checking `is_cancelled()`
```suggestion
        if cancel.is_cancelled() {
            warn!("[gRPC][Indexer] Cancelled, stopping stream");
            return OK(());
        }
```